File: cpp/gas_station.cpp
class gas_station
{
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int n = gas.size();
        
        // Calculate total gas and total cost
        int totalGas = 0;
        int totalCost = 0;
        for (int i = 0; i < n; i++) {
            totalGas += gas[i];
            totalCost += cost[i];
        }
        
        // If total gas is less than total cost, impossible
        if (totalGas < totalCost) {
            return -1;
        }
        
        // Difference array and prefix sum
        vector<int> diff(n);
        for (int i = 0; i < n; i++) {
            diff[i] = gas[i] - cost[i];
        }
        
        // Find minimum prefix sum position
        int minPrefix = 0;
        int currentSum = 0;
        int minIndex = 0;
        
        for (int i = 0; i < n; i++) {
            currentSum += diff[i];
            if (currentSum < minPrefix) {
                minPrefix = currentSum;
                minIndex = i + 1;
            }
        }
        
        return minIndex % n;
    }
};

File: python/gas_station.py
from typing import List

class gas_station:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        n = len(gas)
        
        # Calculate total gas and total cost
        total_gas = sum(gas)
        total_cost = sum(cost)
        
        # If total gas is less than total cost, impossible
        if total_gas < total_cost:
            return -1
        
        # Difference array and prefix sum
        diff = [gas[i] - cost[i] for i in range(n)]
        
        # Find minimum prefix sum position
        min_prefix = 0
        current_sum = 0
        min_index = 0
        
        for i in range(n):
            current_sum += diff[i]
            if current_sum < min_prefix:
                min_prefix = current_sum
                min_index = i + 1
        
        return min_index % n